lib/rules/no-restricted-imports.js

```diff
@@ -234,20 +241,53 @@ module.exports = {
         /**
          * Report a restricted path specifically for patterns.
          * @param {node} node representing the restricted path reference
-         * @param {Object} group contains a Ignore instance for paths, and the customMessage to show if it fails
+         * @param {Object} group contains an Ignore instance for paths, the customMessage to show on failure,
+         * and any restricted import names that have been specified in the config
+         * @param {Map<string,Object[]>} importNames Map of import names that are being imported
          * @returns {void}
          * @private
          */
-        function reportPathForPatterns(node, group) {
+        function reportPathForPatterns(node, group, importNames) {
             const importSource = node.source.value.trim();
 
-            context.report({
-                node,
-                messageId: group.customMessage ? "patternWithCustomMessage" : "patterns",
-                data: {
-                    importSource,
-                    customMessage: group.customMessage
+            const customMessage = group.customMessage;
+            const restrictedImportNames = group.importNames || [];
+
+            /*
+             * If we are not restricting to any specific import names and just the pattern itself,
+             * report the error and move on
+             */
+            if (restrictedImportNames.length < 1) {
+                context.report({
+                    node,
+                    messageId: customMessage ? "patternWithCustomMessage" : "patterns",
+                    data: {
+                        importSource,
+                        customMessage
+                    }
+                });
+                return;
+            }
+
+            restrictedImportNames.forEach(importName => {
+                if (!importNames.has(importName)) {
+                    return;
                 }
+
+                const specifiers = importNames.get(importName);
+
+                specifiers.forEach(specifier => {
+                    context.report({
+                        node,
+                        messageId: customMessage ? "importNameWithCustomMessage" : "importName",
```

Comment:
I think we need two new `messageId`s for this, to note in the message text that this restriction comes from `patterns`, like we already have [`"patterns"`](https://github.com/eslint/eslint/blob/b6aee9591ecc2e2f5738ab8bef20faac1e05b5c3/lib/rules/no-restricted-imports.js#L101) as a different message than the [`"path"`](https://github.com/eslint/eslint/blob/b6aee9591ecc2e2f5738ab8bef20faac1e05b5c3/lib/rules/no-restricted-imports.js#L97) message.